### PR TITLE
Fix drift in train_cartpole*.rs examples (closes #18)

### LINKED ISSUE #18: Examples drift: train_cartpole*.rs has stale TrainingMetadata, SpaceInfo, and buffer.add() signatures
## Background

Same class of bug as #7 (now fixed in main lib code), but in the `examples/games/cartpole/` directory. The lib unit tests now compile, but the cartpole training examples still don't — they were never updated when the underlying types/traits changed. This is breaking the `Tests (Linux + libtorch)` and `Tests (macOS + libtorch)` CI jobs.

## Affected files

- `examples/games/cartpole/train_cartpole.rs`
- `examples/games/cartpole/train_cartpole_best.rs`
- `examples/games/cartpole/train_cartpole_hybrid.rs`
- `examples/games/cartpole/train_cartpole_long.rs`
- `examples/games/cartpole/train_cartpole_modern.rs`
- `examples/games/cartpole/train_cartpole_sb3.rs`

(Audit each — drift may not be uniform.)

## Known breakage modes (per the PR #16 judge review)

- `TrainingMetadata.hyperparameters` — field missing / renamed.
- `SpaceInfo.dtype` — renamed to `space_type` (this is the same rename that #7 fixed in `src/env/games/cartpole.rs`).
- `buffer.add()` — signature drift (now takes 9 args per `src/buffer/rollout/storage.rs`; was different before).

## Reproduction

```bash
cargo check --features training --examples
```

## Acceptance

- `cargo check --features training --examples` succeeds.
- `cargo build --features training --examples` succeeds.
- Tests CI jobs go green (modulo the unrelated rustdoc job — see separate issue).

## Discovery

Flagged by the Loom Judge during re-review of PR #16 ([comment](https://github.com/rjwalters/thrust/pull/16#issuecomment-4455152478)). Out of scope for #7, which was scoped to library unit tests only.

### PR DESCRIPTION
**Pull request by rjwalters**
## Summary

Closes #18.

Five example files in `examples/games/cartpole/` had drifted from the current `master` API surface and failed to compile under `cargo check --features training --examples` (11 compile errors across 5 of the 6 listed files; `train_cartpole_best.rs` already compiled cleanly and is untouched).

This patch reapplies the API renames so the training examples build again.

### Per-file fixes

| File | Errors | What changed |
|---|---|---|
| `train_cartpole.rs` | 2 | `action_space.dtype` -> `.space_type`; `buffer.add` now takes `&observations[env_id]` (was `.clone()`). |
| `train_cartpole_best.rs` | 0 | **Unchanged.** Already correct. |
| `train_cartpole_hybrid.rs` | 1 | Add `hyperparameters: None,` to `TrainingMetadata`. |
| `train_cartpole_long.rs` | 6 | `.dtype`->`.space_type`; `&observations[env_id]`; flat `batch.observations` (no `.flatten().copied()`); `batch_size = batch.observations.len() / obs_dim as usize` (fixes a latent logic bug from the old nested-vec shape); `batch.log_probs`->`batch.old_log_probs`; `batch.values`->`batch.old_values`. |
| `train_cartpole_modern.rs` | 1 | Add `hyperparameters: None,`. |
| `train_cartpole_sb3.rs` | 1 | Add `hyperparameters: None,`. |

### Authoritative shape references

- `TrainingMetadata` — `src/policy/inference.rs:226` (new required field `hyperparameters: Option<HashMap<String, serde_json::Value>>`, no `Default` derived).
- `SpaceInfo` — `src/env/mod.rs:51` (field renamed `dtype` -> `space_type`).
- `RolloutBuffer::add` — `src/buffer/rollout/storage.rs:107` (9-arg signature; `observation: &[f32]`).
- `RolloutBatch` — `src/buffer/rollout/storage.rs:212` (`old_log_probs`, `old_values`; `observations` is a flat `Vec<f32>`).

### Out of scope

`examples/games/cartpole/debug_cartpole.rs` (9 errors) is a different drift class — stale `env.reset()?` / `env.step(action)?` from when `Environment` still returned `Result`. Curator flagged it as out of scope for this issue; it should be tracked in a separate follow-up.

## Test plan

- [x] `cargo check --features training --example train_cartpole` (clean)
- [x] `cargo check --features training --example train_cartpole_best` (clean, untouched file still passes)
- [x] `cargo check --features training --example train_cartpole_hybrid` (clean)
- [x] `cargo check --features training --example train_cartpole_long` (clean)
- [x] `cargo check --features training --example train_cartpole_modern` (clean)
- [x] `cargo check --features training --example train_cartpole_sb3` (clean)
- [x] `cargo build --features training --example <each>` (all six link successfully)
- [x] `cargo +nightly fmt` on changed files
- [ ] CI `Tests (Linux + libtorch)` and `Tests (macOS + libtorch)` go green

### CODE REVIEW

**Comment by rjwalters** [comment]
## Approval — train_cartpole*.rs drift fix

LGTM. The diffs match the curator's punch table exactly, and `cargo check --features training --examples` errors only on `debug_cartpole.rs` (which is explicitly out-of-scope per the PR body and curator scoping).

### Verification

**Per-file diff vs. punch table**:

| File | Punch table | Observed | Match |
|---|---|---|---|
| `train_cartpole.rs` | `.dtype`→`.space_type` + `&observations[env_id]` (2 edits) | confirmed at L38, L110 | yes |
| `train_cartpole_best.rs` | unchanged | empty diff vs. `main` | yes |
| `train_cartpole_hybrid.rs` | `hyperparameters: None,` | confirmed at L259 | yes |
| `train_cartpole_modern.rs` | `hyperparameters: None,` | confirmed at L262 | yes |
| `train_cartpole_sb3.rs` | `hyperparameters: None,` | confirmed at L265 | yes |
| `train_cartpole_long.rs` | 6 edits incl. flat batch, `old_log_probs`, `old_values`, `batch_size = …/obs_dim` | all 6 confirmed at L40, L114, L156–165 | yes |

**Authoritative shape cross-check**:
- `RolloutBatch.observations: Vec<f32>` (flat) — `src/buffer/rollout/storage.rs:214`
- `RolloutBatch.old_log_probs`, `RolloutBatch.old_values` — L220, L223
- `RolloutBuffer::add(observation: &[f32], …)` — L107–117
- `SpaceInfo.space_type: SpaceType` — `src/env/mod.rs:58`
- `TrainingMetadata.hyperparameters: Option<HashMap<…>>` — `src/policy/inference.rs:246`

**Compilation check** (local, `LIBTORCH_USE_PYTORCH=1`):
- `cargo check --features training --example train_cartpole` → clean
- `cargo check --features training --example {train_cartpole_best, _hybrid, _long, _modern, _sb3}` → clean (all five)
- `cargo check --features training --examples` → only `debug_cartpole` fails (9 errors, all of the `?`-on-`reset`/`step` family, pre-existing on `main`)

### Particularly worth noting

- The `batch_size = batch.observations.len() / obs_dim as usize` rewrite in `train_cartpole_long.rs:158` is dimensionally correct given `RolloutBatch.observations` is a flat `Vec<f32>` of length `num_steps * num_envs * obs_dim`. The pre-PR code set `batch_size = batch.observations.len()` (total floats, not transitions), which would have miscomputed the reshape dim and either panicked or produced an incorrectly-shaped tensor. Latent bug, correctly fixed.
- `train_cartpole.rs` already had `old_log_probs`/`old_values` on `main` (from prior work), so the curator was right to scope it to only the 2 remaining edits.

### CI status (informational)

`mergeStateStatus = UNSTABLE`. The failing checks are pre-existing on `main`, not regressions:
- `Tests (Linux + libtorch)` / `Tests (macOS + libtorch)`: fail because `cargo test --features training` builds all examples, and `debug_cartpole.rs` is broken on `main` (stale `env.reset()?` / `env.step(action)?` — different drift class, intentionally deferred per curator scoping; PR body explicitly tracks this as a follow-up).
- `Documentation`: fails on pre-existing `missing-docs` errors throughout the library (`src/optimize/space.rs` etc.), unrelated to this PR.

Approving despite UNSTABLE merge state because (a) all failures are pre-existing baseline noise on `main` that this PR cannot reasonably be asked to fix, (b) the curator explicitly scoped `debug_cartpole.rs` out, and (c) the PR's stated acceptance (`train_cartpole*.rs` compile clean) is satisfied. A follow-up issue for `debug_cartpole.rs` should be filed; that work should also unblock the Tests CI jobs.

Approved.
